### PR TITLE
[PT-D] Add Tensor Parallel _prepare_output component functions

### PR DESCRIPTION
**Pull request by fduwjj**
This is part of efforts trying to standardize and make Tensor Parallel(TP) API Beta Release. 

Since we will use DTensor for all TP internal interactions, we need to convert the output to the state(DTensor / Tensor) which user wants.

### CODE REVIEW

File: spmd/tensor/parallel/utils.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+        device_mesh = args[1]
+        assert isinstance(
+            output, DTensor
+        ), f"output of Tensor Parallel is actually {type(output)} not DTensor."
+        if device_mesh is None:
+            device_mesh = output.device_mesh
+
+        assert (
+            device_mesh.ndim == 1
+        ), f"{_prepare_output_func.__name__}: device mesh is not 1D"
```
Comment (XilunWu) [inline spmd/tensor/parallel/utils.py:48]: I'm thinking that the function name part "{_prepare_output_func.__name__}" is not necessary because it will also be displayed in the error trace. What I have now is:

f"device_mesh dim is {device_mesh.ndim} but expcted to be 1"

What do you think?

File: spmd/tensor/parallel/utils.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    Args:
+        _prepare_output_func (Callable): The func we want to inject the 
+            validation into.
+    Return:
+        func (Callable): Same input func with validation logic added.
+    """
+
+    @functools.wraps(_prepare_output_func)
+    def wrapper(*args, **kwargs):  # pyre-ignore[2, 3]
+        assert len(args) >= 2, "_prepare_output need at least two args."
```
Comment (XilunWu) [inline spmd/tensor/parallel/utils.py:37]: Question:
If the call is `make_output_replicate(tensor)` (i.e. device_mesh argument passed is None), will `len(args)` be just 1?

**Review by XilunWu — APPROVED** [review]
Thx for working on this! I'll adjust my PR to keep a uniform style. Left some comments.

File: spmd/tensor/parallel/utils.py (hunk 2 above)
Comment (fduwjj) [reply spmd/tensor/parallel/utils.py:37]: yep.

File: spmd/tensor/parallel/utils.py (hunk 1 above)
Comment (fduwjj) [reply spmd/tensor/parallel/utils.py:48]: OK, maybe we still want to say it's input or output. Let me remove the func name. Thanks!

File: spmd/tensor/parallel/style.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    """
+    Convert Output DTensor to a replicated DTensor. This will be used in ParallelStyle.
+    Args:
+        output (DTensor): output of module to be converted.
+        device_mesh (Optional[DeviceMesh]): DeviceMesh to replicate the output DTensor.
+            This needs to be a 1D DeviceMesh and we will throw exceptions if a non-1D
+            DeviceMesh is passed in. If no DeviceMesh is passed in, we will reuse the
+            one from output DTensor.
+    Return:
+        (DTensor): A DTensor made repliacted.
```
Comment (wanchaol) [inline spmd/tensor/parallel/style.py:41]: typo: replicate

File: spmd/tensor/parallel/utils.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
+import functools
+
+import torch
+from spmd.tensor import DeviceMesh, DTensor
+from typing import Callable, Optional, Union
+
+
+def _prepare_output_validate(
+    _prepare_output_func: Callable[
+        [DTensor, Optional[DeviceMesh], Optional[int]],
```
Comment (wanchaol) [inline spmd/tensor/parallel/utils.py:10]: can we make this long callable type a type alias?

File: spmd/tensor/parallel/utils.py (hunk 2 above)
Comment (wanchaol) [reply spmd/tensor/parallel/utils.py:37]: if we need at least two args, let's just make device mesh be non-optional?

File: spmd/tensor/parallel/utils.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    """
+
+    @functools.wraps(_prepare_output_func)
+    def wrapper(*args, **kwargs):  # pyre-ignore[2, 3]
+        assert len(args) >= 2, "_prepare_output need at least two args."
+        output = args[0]
+        device_mesh = args[1]
+        assert isinstance(
+            output, DTensor
+        ), f"output of Tensor Parallel is actually {type(output)} not DTensor."
```
Comment (wanchaol) [inline spmd/tensor/parallel/utils.py:42]: nit: expect output of Tensor Parallel be a DTensor, but found ...

**Review by wanchaol — COMMENTED** [review]
looks good overall, some nits

File: spmd/tensor/parallel/utils.py (hunk 4 above)
Comment (fduwjj) [reply spmd/tensor/parallel/utils.py:10]: I see. Sure.

File: spmd/tensor/parallel/style.py (hunk 3 above)
Comment (fduwjj) [reply spmd/tensor/parallel/style.py:41]: opps.

File: spmd/tensor/parallel/utils.py (hunk 2 above)
Comment (fduwjj) [reply spmd/tensor/parallel/utils.py:37]: Opps, I thought it in a wrong way. We do want to make DeviceMesh Optional.. I will change the code.

**Review by wanchaol — APPROVED** [review]

File: spmd/tensor/parallel/utils.py [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    Args:
+        _prepare_output_func (Callable): The func we want to inject the
+            validation into.
+    Return:
+        func (Callable): Same input func with validation logic added.
+    """
+
+    @functools.wraps(_prepare_output_func)
+    def wrapper(*args, **kwargs):  # pyre-ignore[2, 3]
+        assert len(args) >= 1, "_prepare_output need at least one arg."
```
Comment (XilunWu) [inline spmd/tensor/parallel/utils.py:36]: nit: need -> needs

File: spmd/tensor/parallel/utils.py [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+        output = args[0]
+        assert isinstance(
+            output, DTensor
+        ), f"Expect output of Tensor Parallel to be a DTensor, but found {type(output)}."
```
Comment (XilunWu) [inline spmd/tensor/parallel/utils.py:40]: Sorry for the confusion but.

One question:
Upon the exception being thrown, will the trace display the location of error as in the wrapper or in the actual `_prepare_output_func`? If the former, will it still be better to print `_prepare_output_func.__name__`? Sorry I wasn't thinking about this.

Comment (fduwjj) [reply spmd/tensor/parallel/utils.py:40]: I think with `@functools.wraps`, it should give you the func name. You can try and verify that.

File: spmd/tensor/parallel/utils.py [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+        if len(args) < 2:
+            device_mesh = output.device_mesh
+        else:
+            device_mesh = args[1]
+
+        assert (
+            device_mesh.ndim == 1
+        ), f"device_mesh has dims {device_mesh.ndim} but expcted to be 1 for output."
+        return _prepare_output_func(*args, **kwargs)
```
Comment (XilunWu) [inline spmd/tensor/parallel/utils.py:49]: Just found a potential issue: 
In case where `len(args) < 2` (i.e. device_mesh is not passed to func), we need to append it back to `args` before calling `_prepare_output_func(*args, **kwargs)`. Our existing test case did not capture it.

Comment (fduwjj) [reply spmd/tensor/parallel/utils.py:49]: Fixed in https://github.com/pytorch/tau/pull/619.
